clang/lib/Basic/Diagnostic.cpp

```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
       setSeverity(Diag, Map, Loc);
 }
 
+namespace {
+class WarningsSpecialCaseList : public llvm::SpecialCaseList {
```

Comment:
unfortunately we need to use inheritance, as we need access to internals for implemeting some custom "special-case-list" matching logic.

current users of `SpecialCaseList` also go with a similar approach for adding some custom support on top of the parser. I don't feel comfortable changing/exposing all these interfaces directly from `SpecialCaseList`. Especially because it has some backward-compatibility gurantees around regexes vs globs etc, and I don't want to carry that complexity into this new use case.

I think in an ideal world `SpecialCaseList` should just expose its parser, and we'd use it here. I can perform such a refactoring if you feel strong about this.

Reply:
Yes, I think the current state is pretty bad so if you're willing to refactor (in a separate PR, could also be a followup), this would be great. I think WarningsSpecialCaseList::createOrDie() would be surprising and I assume future changes could be harder to make and/or make the API much less clear.

Reply:
Add a FIXME/TODO ?